### PR TITLE
Master auth renewal

### PR DESCRIPTION
**Pull request by kke**
- External OAuth2 providers can be used as authentication provider
- Internal OAuth2 provider now supports refresh token and authorization
  code grant flow
- API endpoint /authenticate is used for performing authentication
- API endpoint /v1/auth no longer available, there is no more username
  and password login
- API endpoint /v1/auth_provider is used to configure external oauth2
  provider settings
- Only encryption hashes are stored for local access and refresh tokens
- Token validation is done in RACK middleware instead of a helper module
- There's a RACK middleware that restricts access to Kontena CLI clients
  prior to version X (currently set to 0.15.0 in server.rb). The error
  message contains a hint about how to upgrade and the old CLI will show
  it when trying to perform requests.

Authentication flow:

Upon deployment a local user "admin" is created. The administrator gets
an access token with a code, generated by the CLI before deployment.

This code is then used by the CLI to perform initial authentication.

The usage of "admin" account does not require that an oauth2 provider is
configured, it's a local only account.

To configure an auth provider you can use the CLI's "kontena master
auth-provider config" command.

Authenticating a regular user:

A) Auth provider not configured:
  1) Client requests /authenticate?redirect_uri=http://localhost:xxx/cb
  2) Master responds with HTTP 500 - Authentication provider not
  configured

B) Auth provider configured
  1) Client requests
     /authenticate?redirect_uri=http://localhost:xxx/cb
     or if the user has an invitation code:
     /authenticate?redirect_uri=..&invite_code=<invite_code>
  2) Master responds with a HTTP 302 redirect to auth provider
  authorization URL.
  3) Client opens the link in a browser
  4) Once the user permits the authorization, the auth provider
  redirects back to master's /cb with an authorization code
  5) Master exchanges this code to an access token from AP's token
  endpoint.
  6) Master requests user info from AP's user info endpoint. Using the
  access token.
  7) Local user is determined by matching the external_id field.
  For first time users an invitation code is required.
  8) Master creates a local access token with an authorization code
  9) Master responds with a redirect to
  http://localhost:xxx/cb?code=<authorization_code>
  10) Client exchanges the authorization code to an access token using
  master's /oauth2/token

The Oauth2 consumer used to authenticate from external providers should
be configurable to work with almost any provider that follows the Oauth2
specification at some level. It requires that you have created an Oauth2
application on the provider and have the client_id and client_secret.

The external access tokens are currently not stored on master as it
serves no purpose because the token is not used after the authentication.
In the future it might be used to sync user data or for example to send
notifications. At that point it would be a good idea to store them in
some encrypted format or in kontena vault.

The server's oauth2 application client secret should also be stored in
such way.

TODO/Questions:
 - how to migrate existing installations?
 - the authentication related things probably need to write something to audit logs?
 - creating service users and tokens for automation and headless operation

### CODE REVIEW

File: server/app/models/access_token.rb [hunk 1]
```diff
@@ -1,18 +1,202 @@
...
+  validates :refresh_token, uniqueness: { scope: :internal, allow_nil: true }
+
+  field :token_type, type: String, default: 'bearer'
   field :token, type: String
   field :refresh_token, type: String
-  field :expires_at, type: Time
+  field :expires_at, type: DateTime
   field :scopes, type: Array
+  field :deleted_at, type: BSON::Timestamp, default: nil
+  field :internal, type: Boolean, default: true
```
Comment (jakolehm) [inline server/app/models/access_token.rb]: What is the purpose of `internal` ?

File: server/app/mutations/users/invite.rb [hunk 2]
```diff
@@ -2,24 +2,50 @@ module Users
   class Invite < Mutations::Command
     required do
       model :user
+    end
+
+    optional do
```
Comment (jakolehm) [inline server/app/mutations/users/invite.rb]: What happens if we don't receive any of the optional params?

File: server/app/routes/v1/audit_logs_api.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    include RequestHelpers
+    include Auditor
+
+    plugin :streaming
+    
+    route do |r|
+      r.get do
+        halt(403, 'Access denied') unless current_user.master_admin?
+
+        limit = request.params['limit'] || 500
```
Comment (jakolehm) [inline server/app/routes/v1/audit_logs_api.rb]: We should check this limit better so that we don't fetch too many items from db.

File: server/app/services/memory_store.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# Simple memory store for Rack::Attack throttling
+class MemoryStore
```
Comment (jakolehm) [inline server/app/services/memory_store.rb]: Not a thread safe implementation. One option is to use [MonitorMixin](http://ruby-doc.org/stdlib-2.3.1/libdoc/monitor/rdoc/MonitorMixin.html) (or plain old mutex).

File: server/app/boot_jobs.rb [hunk 5]
```diff
@@ -5,6 +5,7 @@
 unless ENV['RACK_ENV'] == 'test'
   MongoPubsub.start!(PubsubChannel.collection)
   JobSupervisor.run!
+  Mongodb::Migrator.new.migrate
```
Comment (kke) [inline server/app/boot_jobs.rb]: Puma hooks may not be a good spot if you want to do something like `irb -r ./server.rb`

File: server/app/initializers/symmetric_encryption.rb [hunk 6]
```diff
@@ -1,10 +1,15 @@
 require 'symmetric-encryption'
 require 'symmetric_encryption/extensions/mongoid/encrypted'
 
+if ENV['VAULT_INIT']
```
Comment (kke) [inline server/app/initializers/symmetric_encryption.rb]: To init random vault keys when running locally.

File: server/app/models/access_token.rb (hunk 1 above)
Comment (kke) [inline server/app/models/access_token.rb]: internal: false is for auth provider's tokens. 
internal: true is for authenticating to this master

File: server/app/mutations/users/invite.rb (hunk 2 above)
Comment (kke) [inline server/app/mutations/users/invite.rb]: Then errors are added through model validations:

```
        end_user.errors.each do |key, message|
          add_error(key, :invalid, message)
        end
```

File: server/app/routes/v1/audit_logs_api.rb (hunk 3 above)
Comment (kke) [inline server/app/routes/v1/audit_logs_api.rb]: The thing from container logs that @SpComb made should probably be integrated into the audit logs too. With tailing.

File: server/app/boot_jobs.rb (hunk 5 above)
Comment (kke) [inline server/app/boot_jobs.rb]: Probably most obvious place would be the mongo initializer.

File: server/app/services/memory_store.rb (hunk 4 above)
Comment (kke) [inline server/app/services/memory_store.rb]: Or even easier, [ActiveSupport::Cache::MemoryStore](http://apidock.com/rails/v4.2.7/ActiveSupport/Cache/MemoryStore) since we already have active support in gems.

File: server/app/boot_jobs.rb (hunk 5 above)
Comment (kke) [inline server/app/boot_jobs.rb]: moved to boot.rb

File: server/app/helpers/token_authentication_helper.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  include OAuth2Api::Common
+
+  def validate_access_token(*scopes)
+    # These only happen when in a "soft exclude" path where
+    # the headers are processed but request is not halted
+    # by the token authentication middleware.
+    
+    unless current_user
+      Server.logger.debug "No current user"
```
Comment (kke) [inline server/app/helpers/token_authentication_helper.rb]: fixed by using the existing Logging module for logging purposes.

File: server/app/initializers/symmetric_encryption.rb (hunk 6 above)
Comment (kke) [inline server/app/initializers/symmetric_encryption.rb]: removed. dotenv can be used locally.

File: server/app/boot.rb [hunk 8]
```diff
@@ -30,3 +30,5 @@
 Dir[__dir__ + '/jobs/**/*.rb'].each {|file| require file }
 
 Dir[__dir__ + '/services/**/*.rb'].each {|file| require file }
+
+Mongodb::Migrator.new.migrate
```
Comment (jakolehm) [inline server/app/boot.rb]: Still not a correct place for this.. this problem is already fixed in the master so imho this PR should not try to re-implement it differently? :)

See #999

/cc @jnummelin @nevalla

File: server/app/helpers/token_authentication_helper.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    # the headers are processed but request is not halted
+    # by the token authentication middleware.
+    
+    unless current_user
+      debug "No current user"
+      mime_halt(403, 'access_denied')
+    end
+
+    unless current_access_token
+      Server.logger.debug "No current token"
```
Comment (jakolehm) [inline server/app/helpers/token_authentication_helper.rb]: Maybe just `debug`

File: server/app/helpers/config_helper.rb [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+      @salt
+    end
+
+    def digest(string)
+      BCrypt::Engine.hash_secret(string, salt)
+    end
+
+    def config
+      return @config if @config
+      if Object::const_defined?('Server')
```
Comment (jakolehm) [inline server/app/helpers/config_helper.rb]: Why we need this if/elsif/else ?

File: server/app/routes/v1/audit_logs_api.rb (hunk 3 above)
Comment (jakolehm) [inline server/app/routes/v1/audit_logs_api.rb]: Yes, but at least we need to plug this "bug" so that server won't crash if too many logs are requested? Refactor to @SpComb implementation should be in a separate PR and not here.

File: server/db/migrations/13_recreate_access_tokens.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+class RecreateAccessTokens < Mongodb::Migration
+  def self.up
+    AccessToken.collection.drop
```
Comment (jakolehm) [inline server/db/migrations/13_recreate_access_tokens.rb]: Is there any sane way to migrate old tokens to new ones? If yes, we should implement that in a separate PR. This is ok implementation for this PR :)

File: server/server.rb [hunk 12]
```diff
@@ -13,7 +19,82 @@ class Server < Roda
...
+  def self.root_url
+    @url ||= config[:root_url]
+  end
+
+  def self.name
+    @name ||= config[:server_name]
+  end
+
+  # Global logger
+  def self.logger
```
Comment (jakolehm) [inline server/server.rb]: Why not include `Logging` module? This seems a little bit duplicate implementation to me.

Comment (kke) [inline server/server.rb]: Yes, that is no longer used and is now removed.

File: server/app/boot.rb (hunk 8 above)
Comment (nevalla) [inline server/app/boot.rb]: I agree, let's rely on fix on master branch on this.

File: server/app/helpers/token_authentication_helper.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  include Logging
+
+  def validate_access_token(*scopes)
+    # These only happen when in a "soft exclude" path where
+    # the headers are processed but request is not halted
+    # by the token authentication middleware.
+    
+    unless current_user
+      debug "No current user"
+      mime_halt(403, 'access_denied')
```
Comment (nevalla) [inline server/app/helpers/token_authentication_helper.rb]: How this `mime_halt` differs from `halt_request`?

File: server/db/migrations/13_recreate_access_tokens.rb (hunk 11 above)
Comment (kke) [inline server/db/migrations/13_recreate_access_tokens.rb]: Yes. 

Something like:

```
AccessToken.each |at|
  unless at.token.start_with?('$')
    at.token_plain = at.token
    at.refresh_token = nil
    at.expires_at = nil
    at.internal = true
    at.save
  end
end
```

File: server/app/mutations/users/invite.rb [hunk 14]
```diff
@@ -2,24 +2,50 @@ module Users
...
+      string :grid
```
Comment (nevalla) [inline server/app/mutations/users/invite.rb]: I think this grid option is redundant here(?)

File: server/app/routes/oauth2/authenticate_api.rb [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+        redirect_uri = URI.parse(params['redirect_uri'])
+
+        # Only allow redirect to client's localhost or back to same server's /code'
+        unless (redirect_uri.host && redirect_uri.host == 'localhost') || (redirect_uri.host.nil? && redirect_uri.path == '/code')
+          mime_halt(400, 'invalid_request', 'Invalid redirect_uri') and return
+        end
+
+        invite_code = params['invite_code']
+        if invite_code.kind_of?(String) && invite_code.length > 4
```
Comment (nevalla) [inline server/app/routes/oauth2/authenticate_api.rb]: What happens if `invite_code` is less than 5 chars?

File: server/app/helpers/config_helper.rb (hunk 10 above)
Comment (kke) [inline server/app/helpers/config_helper.rb]: We don't. Overgeneralization.

File: server/app/routes/oauth2/callback_api.rb [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+  # the AP.
+  #
+  # If the AuthorizationRequest object includes a redirect uri
+  # then the final step will redirect the browser to that uri.
+  class CallbackApi < Roda
+    include RequestHelpers
+    include Logging
+
+    route do |r|
+      r.get do
```
Comment (nevalla) [inline server/app/routes/oauth2/callback_api.rb]: This is quite a huge route. Any chance to refactor this into smaller pieces, maybe use some helper methods?

File: server/app/routes/v1/auth_provider.rb [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+        @auth_provider = AuthProvider.instance
+        render('auth_provider/show')
+      end
+
+      r.post do
+        @auth_provider = AuthProvider.instance
+
+        validate_access_token('master_admin') unless current_user_admin?
+
+        halt_request(403, 'Access denied. You must be the local administrator.') unless current_user.email == 'admin'
```
Comment (nevalla) [inline server/app/routes/v1/auth_provider.rb]: How about implementing `User#internal_admin?` (or similar) method instead of comparing email here?

File: server/app/boot.rb (hunk 8 above)
Comment (kke) [inline server/app/boot.rb]: #999 doesn't run migrations except when running puma, it's not running them before tests either.

File: server/app/routes/v1/audit_logs_api.rb (hunk 3 above)
Comment (kke) [inline server/app/routes/v1/audit_logs_api.rb]: done, limited to max 3000.

File: server/app/helpers/token_authentication_helper.rb (hunk 13 above)
Comment (kke) [inline server/app/helpers/token_authentication_helper.rb]: It responds with correct mime type and format and adds [WWW-Authenticate header](http://self-issued.info/docs/draft-ietf-oauth-v2-bearer.html#authn-header).

File: server/app/routes/oauth2/authenticate_api.rb (hunk 15 above)
Comment (kke) [inline server/app/routes/oauth2/authenticate_api.rb]: The code is ignored. The callback will try to find the user by using email. But it should probably halt if you try an invalid code. Fixed now.

File: server/app/routes/v1/auth_provider.rb (hunk 17 above)
Comment (kke) [inline server/app/routes/v1/auth_provider.rb]: yes.

Comment (kke) [inline server/app/routes/v1/auth_provider.rb]: there already is `is_local_admin?`

File: server/app/mutations/users/invite.rb (hunk 14 above)
Comment (kke) [inline server/app/mutations/users/invite.rb]: Correct, removed.

**Comment by kke** [comment]
Included in #1035
